sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/literals.scala

```diff
@@ -46,12 +57,17 @@ object Literal {
     case s: String => Literal(UTF8String.fromString(s), StringType)
     case b: Boolean => Literal(b, BooleanType)
     case d: BigDecimal => Literal(Decimal(d), DecimalType(Math.max(d.precision, d.scale), d.scale))
-    case d: java.math.BigDecimal =>
+    case d: JavaBigDecimal =>
       Literal(Decimal(d), DecimalType(Math.max(d.precision, d.scale), d.scale()))
     case d: Decimal => Literal(d, DecimalType(Math.max(d.precision, d.scale), d.scale))
     case t: Timestamp => Literal(DateTimeUtils.fromJavaTimestamp(t), TimestampType)
     case d: Date => Literal(DateTimeUtils.fromJavaDate(d), DateType)
     case a: Array[Byte] => Literal(a, BinaryType)
+    case a: Array[_] =>
```

Comment:
Seems your code only works at compile time.
Actually, it did not work in runtime.
I checked;
```
    case a: Array[_] =>
      val classSymbol = ru.runtimeMirror(getClass.getClassLoader).classSymbol(a.getClass)
      val tpe = classSymbol.selfType
      ScalaReflection.schemaFor(tpe).dataType
```
This threw an exception because type signatures were not available in runtime;
```
  java.lang.UnsupportedOperationException: Schema for type T is not supported
  at org.apache.spark.sql.catalyst.ScalaReflection$.schemaFor(ScalaReflection.scala:704)
  at org.apache.spark.sql.catalyst.ScalaReflection$.schemaFor(ScalaReflection.scala:660)
  at org.apache.spark.sql.catalyst.expressions.Literal$.apply(literals.scala:73)
```